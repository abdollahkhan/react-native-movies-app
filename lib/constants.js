export const API_URL = 'https://api.themoviedb.org/3'
export const API_KEY = '0511c9bf8a6b1d1216961e19831d9a40'

export const SCREENS = {
  SETTINGS: 'Settings',
  FAVOURITES: 'Favourites',
  DISCOVER: 'Discover',
  MAIN: 'Main'
}
