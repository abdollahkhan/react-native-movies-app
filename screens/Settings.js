import {
  KeyboardAvoidingView,
  Platform,
  ScrollView,
  StyleSheet,
  Text,
  TextInput,
  TouchableOpacity,
  View
} from 'react-native'
import { useSafeAreaInsets } from 'react-native-safe-area-context'
import { useDispatch, useSelector } from 'react-redux'
import { Genre, SortOption } from '../components'
import { SORT_OPTIONS } from '../lib/constants'

import {
  colors,
  containerStyles,
  fontWeights,
  sizes,
  textStyles
} from '../lib/styles'
import {
  addSortFilter,
  addYearFilter,
  removeGenreFilter,
  removeSortFilter,
  addGenreFilter,
  addRuntimeFromFilter,
  addRuntimeToFilter
} from '../redux/reducers/filters.reducer'
import {
  selectRuntimeFromFilter,
  selectRuntimeToFilter,
  selectSortFilter,
  selectYearFilter
} from '../redux/selectors/filter.selector'
import { selectAllgenre } from '../redux/selectors/genre.selector'
import { fetchMovies } from '../redux/thunks/movies.thunk'

export default function Settings({ navigation }) {
  const insets = useSafeAreaInsets()
  const dispatch = useDispatch()

  const year = useSelector(selectYearFilter)
  const runtimeFrom = useSelector(selectRuntimeFromFilter)
  const runtimeTo = useSelector(selectRuntimeToFilter)

  return (
    <View
      style={[
        containerStyles,
        {
          paddingBottom: insets.bottom
        }
      ]}
    >
      <KeyboardAvoidingView
        behavior={Platform.OS === 'ios' ? 'padding' : 'height'}
        style={styles.keyboardView}
      >
        <ScrollView contentContainerStyle={styles.wrapper}>
          <View>
            <Text style={textStyles.h2}>Sort by</Text>
            <View>
              <SortOptions />
            </View>
          </View>
          <View>
            <Text style={textStyles.h2}>Genres</Text>
            <View style={styles.genreList}>
              <Genres />
            </View>
          </View>
          <View>
            <Text style={textStyles.h2}>Year</Text>
            <TextInput
              keyboardType="number-pad"
              style={styles.input}
              maxLength={4}
              value={year}
              onChangeText={(val) => dispatch(addYearFilter(val))}
            />
          </View>
          <View>
            <Text style={textStyles.h2}>Runtime</Text>
            <View style={styles.runtime}>
              <TextInput
                keyboardType="number-pad"
                style={styles.input}
                placeholder="From"
                placeholderTextColor={colors.neutral}
                maxLength={3}
                value={runtimeFrom}
                onChangeText={(val) => dispatch(addRuntimeFromFilter(val))}
              />
              <Text style={textStyles.small}>-</Text>
              <TextInput
                keyboardType="number-pad"
                style={styles.input}
                placeholder="To"
                placeholderTextColor={colors.neutral}
                maxLength={3}
                value={runtimeTo}
                onChangeText={(val) => dispatch(addRuntimeToFilter(val))}
              />
              <Text style={textStyles.small}>minutes</Text>
            </View>
          </View>
        </ScrollView>
        <View style={styles.buttonContainer}>
          <TouchableOpacity
            activeOpacity={0.7}
            underlayColor={colors.neutral}
            style={styles.button}
            onPress={() => {
              dispatch(fetchMovies())
              navigation.goBack()
            }}
          >
            <Text style={styles.buttonText}>Save</Text>
          </TouchableOpacity>
        </View>
      </KeyboardAvoidingView>
    </View>
  )
}

const SortOptions = () => {
  const dispatch = useDispatch()
  const sortFilter = useSelector(selectSortFilter)

  return Object.keys(SORT_OPTIONS).map((item) => {
    return (
      <SortOption
        key={item}
        name={item}
        selected={item === sortFilter}
        onSelect={() =>
          SORT_OPTIONS[item] !== sortFilter && item !== sortFilter
            ? dispatch(addSortFilter(item))
            : dispatch(removeSortFilter())
        }
      />
    )
  })
}

const Genres = () => {
  const dispatch = useDispatch()

  const genres = useSelector(selectAllgenre)

  const toggleGenreSelect = (id, isSelected) => {
    isSelected ? dispatch(removeGenreFilter(id)) : dispatch(addGenreFilter(id))
  }

  const handleSelect = (id, isSelected) => toggleGenreSelect(id, isSelected)

  return Object.values(genres)?.map((genre) => {
    return (
      <Genre
        key={genre.id}
        id={genre.id}
        name={genre.name}
        onSelect={handleSelect}
      />
    )
  })
}

const styles = StyleSheet.create({
  wrapper: {
    paddingHorizontal: 30,
    paddingVertical: 30,
    gap: 30
  },

  input: {
    backgroundColor: colors.light,
    paddingHorizontal: 12,
    paddingVertical: 10,
    borderRadius: 8,
    fontSize: 14,
    width: 80
  },

  genreList: {
    flexDirection: 'row',
    flexWrap: 'wrap',
    gap: 10
  },

  genre: {
    flexDirection: 'row',
    alignItems: 'center',
    gap: 3,
    borderColor: colors.black,
    borderWidth: 1,
    borderRadius: 8,
    paddingHorizontal: 12,
    paddingVertical: 8,
    justifyContent: 'center'
  },

  runtime: {
    flexDirection: 'row',
    alignItems: 'center',
    gap: 16
  },

  buttonContainer: {
    backgroundColor: colors.white,
    borderTopWidth: 1,
    borderTopColor: colors.light,
    paddingVertical: 16,
    paddingHorizontal: 30
  },

  button: {
    backgroundColor: colors.black,
    paddingHorizontal: 16,
    paddingVertical: 16,
    borderRadius: 16,
    alignItems: 'center'
  },

  buttonText: {
    color: colors.white,
    fontWeight: fontWeights.bold,
    fontSize: sizes.md
  },

  keyboardView: {
    flex: 1
  }
})
