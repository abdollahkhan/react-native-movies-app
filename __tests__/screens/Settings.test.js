import React from 'react'
import { configureStore } from '@reduxjs/toolkit'
import movieReducer from '../../redux/reducers/movie.reducer'
import mediaReducer from '../../redux/reducers/media.reducer'
import { Provider } from 'react-redux'
import {
  render,
  screen,
  fireEvent,
  cleanup
} from '@testing-library/react-native'
import { SORT_OPTIONS } from '../../lib/constants'

import filterReducer from '../../redux/reducers/filters.reducer'
import Settings from '../../screens/Settings'
import { SafeAreaProvider } from 'react-native-safe-area-context'
import genreReducer from '../../redux/reducers/genre.reducer'
import * as movieThunks from '../../redux/thunks/movies.thunk'

jest.mock('../../http-common')

jest.mock('react-native-safe-area-context', () => {
  const inset = {
    top: 0,
    right: 0,
    bottom: 0,
    left: 0
  }
  return {
    SafeAreaProvider: ({ children }) => children,
    SafeAreaConsumer: ({ children }) => children(inset),
    useSafeAreaInsets: () => inset
  }
})

describe('Settings Screen', () => {
  const getStore = (initialState) => {
    const store = configureStore({
      reducer: {
        movies: movieReducer,
        media: mediaReducer,
        filters: filterReducer,
        genres: genreReducer
      },
      preloadedState: initialState
    })

    return store
  }

  afterEach(() => {
    cleanup()
  })

  afterAll(() => {
    jest.restoreAllMocks()
  })

  const mockNavigation = {
    navigate: jest.fn(),
    goBack: jest.fn()
  }

  it('renders correctly', async () => {
    const mockStoreState = {}
    const store = getStore(mockStoreState)

    const tree = render(
      <SafeAreaProvider>
        <Provider store={store}>
          <Settings navigation={mockNavigation} />
        </Provider>
      </SafeAreaProvider>
    )

    expect(tree.toJSON()).toMatchSnapshot()
  })

  it('display filters correctly', async () => {
    const mockStoreState = {
      genres: {
        data: {
          1: {
            id: 1,
            name: 'Drama'
          },
          2: {
            id: 2,
            name: 'Drama'
          },
          3: {
            id: 3,
            name: 'Drama'
          }
        }
      }
    }
    const store = getStore(mockStoreState)

    render(
      <SafeAreaProvider>
        <Provider store={store}>
          <Settings navigation={mockNavigation} />
        </Provider>
      </SafeAreaProvider>
    )

    expect(await screen.findAllByText('Drama')).toHaveLength(3)
    expect(await screen.findByText('Genres')).toBeOnTheScreen()
    expect(await screen.findByText('Sort by')).toBeOnTheScreen()
    expect(await screen.findByText('Year')).toBeOnTheScreen()
    expect(await screen.findByText('Runtime')).toBeOnTheScreen()

    Object.keys(SORT_OPTIONS).forEach(async (item) => {
      expect(await screen.findByText(item)).toBeOnTheScreen()
    })
  })

  it('should display text correctly in "Runtime from" input', async () => {
    const mockStoreState = {
      movies: {
        data: {
          results: []
        }
      }
    }

    const store = getStore(mockStoreState)

    render(
      <SafeAreaProvider>
        <Provider store={store}>
          <Settings navigation={mockNavigation} />
        </Provider>
      </SafeAreaProvider>
    )

    const input = await screen.findByPlaceholderText('From')
    const changedText = '122'

    expect(input).toBeOnTheScreen()

    fireEvent.changeText(input, changedText)

    expect(input.props.value).toBe(changedText)
  })

  it('should display text correctly in "Runtime to" input', async () => {
    const mockStoreState = {
      movies: {
        data: {
          results: []
        }
      }
    }

    const store = getStore(mockStoreState)

    render(
      <SafeAreaProvider>
        <Provider store={store}>
          <Settings navigation={mockNavigation} />
        </Provider>
      </SafeAreaProvider>
    )

    const input = await screen.findByPlaceholderText('To')
    const changedText = '876'

    expect(input).toBeOnTheScreen()

    fireEvent.changeText(input, changedText)

    expect(input.props.value).toBe(changedText)
  })

  it('should display text correctly in "Year" input', async () => {
    const mockStoreState = {
      movies: {
        data: {
          results: []
        }
      }
    }

    const store = getStore(mockStoreState)

    render(
      <SafeAreaProvider>
        <Provider store={store}>
          <Settings navigation={mockNavigation} />
        </Provider>
      </SafeAreaProvider>
    )

    const input = await screen.findByLabelText('Release Year')
    const changedText = '2015'

    expect(input).toBeOnTheScreen()

    fireEvent.changeText(input, changedText)

    expect(input.props.value).toBe(changedText)
  })

  it('should apply filters and navigate back', async () => {
    const mockStoreState = {
      movies: {
        data: {
          results: []
        }
      }
    }

    const store = getStore(mockStoreState)

    jest.spyOn(movieThunks, 'fetchMovies')

    render(
      <SafeAreaProvider>
        <Provider store={store}>
          <Settings navigation={mockNavigation} />
        </Provider>
      </SafeAreaProvider>
    )

    const btn = await screen.findByText('Save')

    fireEvent.press(btn)

    expect(mockNavigation.goBack).toBeCalled()
    expect(movieThunks.fetchMovies).toBeCalled()
  })

  it('should toggle genre filter select', async () => {
    const mockStoreState = {
      genres: {
        data: {
          1: {
            id: 1,
            name: 'Drama'
          },
          2: {
            id: 2,
            name: 'Action'
          },
          3: {
            id: 3,
            name: 'Comedy'
          }
        }
      }
    }

    const store = getStore(mockStoreState)

    render(
      <SafeAreaProvider>
        <Provider store={store}>
          <Settings navigation={mockNavigation} />
        </Provider>
      </SafeAreaProvider>
    )
    const btn1 = await screen.findByText('Action')
    const btn2 = await screen.findByText('Comedy')

    fireEvent.press(btn1)
    fireEvent.press(btn2)

    const genres = Object.values(store.getState().filters.data.genre)
    expect(genres).toHaveLength(2)

    fireEvent.press(btn1)

    const updatedGenres = Object.values(store.getState().filters.data.genre)

    expect(updatedGenres).toHaveLength(1)
  })

  it('should toggle sort filter select', async () => {
    const mockStoreState = {
      filters: {
        data: {
          sort: '',
          runtime: {
            to: '',
            from: ''
          }
        }
      }
    }

    const store = getStore(mockStoreState)

    render(
      <SafeAreaProvider>
        <Provider store={store}>
          <Settings navigation={mockNavigation} />
        </Provider>
      </SafeAreaProvider>
    )
    const btn1 = await screen.findByText('Popularity')
    const btn2 = await screen.findByText('Newest First')

    fireEvent.press(btn1)
    let sort = store.getState().filters.data.sort
    expect(sort).toBe('Popularity')

    fireEvent.press(btn2)
    sort = store.getState().filters.data.sort
    expect(sort).toBe('Newest First')

    fireEvent.press(btn2)
    sort = store.getState().filters.data.sort
    expect(sort).toBeFalsy()
  })
})
