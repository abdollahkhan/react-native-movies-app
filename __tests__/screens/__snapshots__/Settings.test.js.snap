// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Settings Screen renders correctly 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#ffffff",
        "flex": 1,
      },
      {
        "paddingBottom": 0,
      },
    ]
  }
>
  <View
    onLayout={[Function]}
    style={
      [
        {
          "flex": 1,
        },
        {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <RCTScrollView
      contentContainerStyle={
        {
          "gap": 30,
          "paddingHorizontal": 30,
          "paddingVertical": 30,
        }
      }
    >
      <View>
        <View>
          <Text
            style={
              {
                "color": "#000000",
                "fontSize": 18,
                "fontWeight": "700",
                "marginBottom": 16,
              }
            }
          >
            Sort by
          </Text>
          <View>
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "alignItems": "center",
                  "borderBottomColor": "#f0f0f0",
                  "borderBottomWidth": 1,
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "opacity": 1,
                  "paddingVertical": 10,
                }
              }
            >
              <Text
                accessibilityRole="text"
                style={
                  {
                    "color": "#242424",
                  }
                }
              >
                Popularity
              </Text>
              <Text
                allowFontScaling={false}
                selectable={false}
                style={
                  [
                    {
                      "color": "#000000",
                      "fontSize": 24,
                    },
                    undefined,
                    {
                      "fontFamily": "Ionicons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    {},
                  ]
                }
              >
                
              </Text>
            </View>
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "alignItems": "center",
                  "borderBottomColor": "#f0f0f0",
                  "borderBottomWidth": 1,
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "opacity": 1,
                  "paddingVertical": 10,
                }
              }
            >
              <Text
                accessibilityRole="text"
                style={
                  {
                    "color": "#242424",
                  }
                }
              >
                Rating
              </Text>
              <Text
                allowFontScaling={false}
                selectable={false}
                style={
                  [
                    {
                      "color": "#000000",
                      "fontSize": 24,
                    },
                    undefined,
                    {
                      "fontFamily": "Ionicons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    {},
                  ]
                }
              >
                
              </Text>
            </View>
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "alignItems": "center",
                  "borderBottomColor": "#f0f0f0",
                  "borderBottomWidth": 1,
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "opacity": 1,
                  "paddingVertical": 10,
                }
              }
            >
              <Text
                accessibilityRole="text"
                style={
                  {
                    "color": "#242424",
                  }
                }
              >
                Newest First
              </Text>
              <Text
                allowFontScaling={false}
                selectable={false}
                style={
                  [
                    {
                      "color": "#000000",
                      "fontSize": 24,
                    },
                    undefined,
                    {
                      "fontFamily": "Ionicons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    {},
                  ]
                }
              >
                
              </Text>
            </View>
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "alignItems": "center",
                  "borderBottomColor": "#f0f0f0",
                  "borderBottomWidth": 1,
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "opacity": 1,
                  "paddingVertical": 10,
                }
              }
            >
              <Text
                accessibilityRole="text"
                style={
                  {
                    "color": "#242424",
                  }
                }
              >
                Oldest First
              </Text>
              <Text
                allowFontScaling={false}
                selectable={false}
                style={
                  [
                    {
                      "color": "#000000",
                      "fontSize": 24,
                    },
                    undefined,
                    {
                      "fontFamily": "Ionicons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    {},
                  ]
                }
              >
                
              </Text>
            </View>
          </View>
        </View>
        <View>
          <Text
            style={
              {
                "color": "#000000",
                "fontSize": 18,
                "fontWeight": "700",
                "marginBottom": 16,
              }
            }
          >
            Genres
          </Text>
          <View
            style={
              {
                "flexDirection": "row",
                "flexWrap": "wrap",
                "gap": 10,
              }
            }
          />
        </View>
        <View>
          <Text
            style={
              {
                "color": "#000000",
                "fontSize": 18,
                "fontWeight": "700",
                "marginBottom": 16,
              }
            }
          >
            Year
          </Text>
          <TextInput
            accessibilityLabel="Release Year"
            keyboardType="number-pad"
            maxLength={4}
            onChangeText={[Function]}
            style={
              {
                "backgroundColor": "#f0f0f0",
                "borderRadius": 8,
                "fontSize": 14,
                "paddingHorizontal": 12,
                "paddingVertical": 10,
                "width": 80,
              }
            }
            value=""
          />
        </View>
        <View>
          <Text
            style={
              {
                "color": "#000000",
                "fontSize": 18,
                "fontWeight": "700",
                "marginBottom": 16,
              }
            }
          >
            Runtime
          </Text>
          <View
            style={
              {
                "alignItems": "center",
                "flexDirection": "row",
                "gap": 16,
              }
            }
          >
            <TextInput
              keyboardType="number-pad"
              maxLength={3}
              onChangeText={[Function]}
              placeholder="From"
              placeholderTextColor="#757575"
              style={
                {
                  "backgroundColor": "#f0f0f0",
                  "borderRadius": 8,
                  "fontSize": 14,
                  "paddingHorizontal": 12,
                  "paddingVertical": 10,
                  "width": 80,
                }
              }
              value=""
            />
            <Text
              style={
                {
                  "color": "#757575",
                  "fontSize": 12,
                }
              }
            >
              -
            </Text>
            <TextInput
              keyboardType="number-pad"
              maxLength={3}
              onChangeText={[Function]}
              placeholder="To"
              placeholderTextColor="#757575"
              style={
                {
                  "backgroundColor": "#f0f0f0",
                  "borderRadius": 8,
                  "fontSize": 14,
                  "paddingHorizontal": 12,
                  "paddingVertical": 10,
                  "width": 80,
                }
              }
              value=""
            />
            <Text
              style={
                {
                  "color": "#757575",
                  "fontSize": 12,
                }
              }
            >
              minutes
            </Text>
          </View>
        </View>
      </View>
    </RCTScrollView>
    <View
      style={
        {
          "backgroundColor": "#ffffff",
          "borderTopColor": "#f0f0f0",
          "borderTopWidth": 1,
          "paddingHorizontal": 30,
          "paddingVertical": 16,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#000000",
            "borderRadius": 16,
            "opacity": 1,
            "paddingHorizontal": 16,
            "paddingVertical": 16,
          }
        }
      >
        <Text
          style={
            {
              "color": "#ffffff",
              "fontSize": 14,
              "fontWeight": "700",
            }
          }
        >
          Save
        </Text>
      </View>
    </View>
  </View>
</View>
`;
